nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_constructor.rs

```diff
@@ -60,47 +82,293 @@ impl BuiltinGetter for ArrayGetSpecies {
+                let Some(next) = iterator_step_value(agent, &iterator_record)? else {
+                    // iv. If next is done, then
+                    // 1. Perform ? Set(A, "length", 𝔽(k), true).
+                    set(
+                        agent,
+                        a,
+                        PropertyKey::from(BUILTIN_STRING_MEMORY.length),
+                        fk,
+                        true,
+                    )?;
+
+                    // 2. Return A.
+                    return Ok(a.into_value());
+                };
+
+                // v. If mapping is true, then
+                let mapped_value = if mapping {
+                    // 1. Let mappedValue be Completion(Call(mapfn, thisArg, « next, 𝔽(k) »)).
+                    let mapped_value =
+                        call(agent, mapfn, this_arg, Some(ArgumentsList(&[next, fk])));
+
+                    // 2. IfAbruptCloseIterator(mappedValue, iteratorRecord).
+                    let _ = if_abrupt_close_iterator(agent, mapped_value, &iterator_record);
+
+                    mapped_value.unwrap()
+                } else {
+                    // vi. Else,
+                    // 1. Let mappedValue be next.
+                    next
+                };
+
+                // vii. Let defineStatus be Completion(CreateDataPropertyOrThrow(A, Pk, mappedValue)).
+                let define_status = create_data_property_or_throw(agent, a, pk, mapped_value);
+
+                // viii. IfAbruptCloseIterator(defineStatus, iteratorRecord).
+                let _ = if_abrupt_close_iterator(
+                    agent,
+                    define_status.map(|_| Value::Undefined),
+                    &iterator_record,
+                );
+
+                // ix. Set k to k + 1.
+                k += 1;
+            }
+        }
+
+        // 6. NOTE: items is not an Iterable so assume it is an array-like object.
+        // 7. Let arrayLike be ! ToObject(items).
+        let array_like = to_object(agent, items).unwrap();
+
+        // 8. Let len be ? LengthOfArrayLike(arrayLike).
+        let len = length_of_array_like(agent, array_like)? as usize;
+        let len_value = Value::try_from(len as f64).unwrap();
+
+        // 9. If IsConstructor(C) is true, then
+        let a = if is_constructor(agent, this_value) {
+            // SAFETY: Only Functions can be constructors
+            let constructor = unsafe { Function::try_from(this_value).unwrap_unchecked() };
+
+            // a. Let A be ? Construct(C, « 𝔽(len) »).
+            construct(agent, constructor, Some(ArgumentsList(&[len_value])), None)?
+        } else {
+            // 10. Else,
+            // a. Let A be ? ArrayCreate(len).
+            array_create(agent, len, len, None)?.into_object()
+        };
+
+        // 11. Let k be 0.
+        let mut k = 0;
+
+        // 12. Repeat, while k < len,
+        while k < len {
+            // 𝔽(k)
+            let fk = Value::try_from(k as f64).unwrap();
```

Comment:
**issue:** Same stuff here as above, let's keep using integers and not do the string conversion for property keys.